nameresolution/mdns/mdns.go

```diff
@@ -122,49 +129,144 @@ func (a *addressList) next() *string {
+}
+
+type Subscriber struct {
+	AddrChan chan string
+	ErrChan  chan error
+}
+
+func (s *Subscriber) Close() {
+	close(s.AddrChan)
+	close(s.ErrChan)
+}
+
+func NewSubscriber() Subscriber {
+	return Subscriber{
+		// ID is assigned by the pool.
+		AddrChan: make(chan string, 1),
+		ErrChan:  make(chan error, 1),
+	}
+}
+
 // NewResolver creates the instance of mDNS name resolver.
-func NewResolver(logger logger.Logger) nameresolution.Resolver {
-	r := &resolver{
+func NewResolver(logger logger.Logger) *Resolver {
+	r := &Resolver{
+		subs:             make(map[string]*SubscriberPool),
 		appAddressesIPv4: make(map[string]*addressList),
 		appAddressesIPv6: make(map[string]*addressList),
-		refreshChan:      make(chan string),
-		logger:           logger,
+		// an app id for every app that is resolved can be pushed
+		// onto this channel. We don't want to block the sender as
+		// they are resolving the app id as part of the service invocation.
+		// Instead we use a buffered channel to balance the load whilst
+		// the background refreshes are being performed. Once this buffer
+		// becomes full, the sender will block and service invocation
+		// will be delayed. We don't expect a single app to resolve
+		// too many other app IDs so we set this to a sensible value
+		// to avoid over allocating the buffer.
+		refreshChan: make(chan string, 36),
+		// registrations channel to signal the resolver to
+		// stop serving queries for registered app ids.
+		registrations: make(map[string]chan struct{}),
+		// shutdownRefresh channel to signal to stop on-demand refreshes.
+		shutdownRefresh: make(chan struct{}, 1),
+		// shutdownRefreshPeriodic channel to signal to stop periodic refreshes.
+		shutdownRefreshPeridoic: make(chan struct{}, 1),
+		logger:                  logger,
 	}
 
 	// refresh app addresses on demand.
 	go func() {
-		for appID := range r.refreshChan {
-			if err := r.refreshApp(context.Background(), appID); err != nil {
-				r.logger.Warnf(err.Error())
+		for {
+			refreshCtx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			select {
+			case appID := <-r.refreshChan:
+				if err := r.refreshApp(refreshCtx, appID); err != nil {
+					r.logger.Warnf(err.Error())
+				}
+			case <-r.shutdownRefresh:
+				r.logger.Debug("stopping on demand cache refreshes.")
+				return
 			}
 		}
 	}()
 
 	// refresh all app addresses periodically.
 	go func() {
 		for {
-			time.Sleep(refreshInterval)
+			refreshCtx, cancel := context.WithCancel(context.Background())
+			defer cancel()
 
-			if err := r.refreshAllApps(context.Background()); err != nil {
-				r.logger.Warnf(err.Error())
+			select {
+			case <-time.After(refreshInterval):
+				if err := r.refreshAllApps(refreshCtx); err != nil {
+					r.logger.Warnf(err.Error())
+				}
+			case <-r.shutdownRefreshPeridoic:
+				r.logger.Debug("stopping periodic cache refreshes.")
+				return
 			}
 		}
 	}()
 
 	return r
 }
 
-type resolver struct {
+type Resolver struct {
```

Comment:
does this struct need to be exported if only NewResolver is used to access it ?

Reply:
Not exporting it means NewResolver returns an unexported type or we return an interface - both which are linting issues as they hide the concrete type.